src/test/java/org/junitpioneer/jupiter/RestoreEnvironmentVariablesTests.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2016-2022 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.parallel.ExecutionMode.SAME_THREAD;
+
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.MethodOrderer;
+import org.junit.jupiter.api.MethodOrderer.OrderAnnotation;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.parallel.Execution;
+
+/**
+ * Verify proper behavior when annotated on a top level class
+ */
+@DisplayName("RestoreEnvironmentVariables Annotation")
+@ExtendWith(RestoreEnvironmentVariablesTests.VerifyEnvVarsExtension.class) // 1st: Order is important here
+@RestoreEnvironmentVariables
+@TestMethodOrder(OrderAnnotation.class)
+@Execution(SAME_THREAD) // Single thread.  See VerifyEnvVarsExtension inner class
+class RestoreEnvironmentVariablesTests {
```

Comment:
I'm not completely sure on the answer here.  The part I cannot address in the existing `EnvironmentVariableExtensionTests` and `SystemPropertiesExtensionTests` is the case where there are no inner classes / potential oddities introduced by added test scaffolding.  I want to be confident on that use case because I think it is most people's use case (I rarely see people use inner classes to group tests - though perhaps more people should).

These separate test classes were the first ones I wrote to verify the functionality.  With the added tests in the other classes, I don't think they cover new ground other than testing the 'normal usage' use case.

Reply:
This is workable.  In the interest of moving forward, I'm going to call this resolved for now.